presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java

```diff
@@ -0,0 +1,423 @@
+    {
+        for (double value : values) {
+            if (!isNaN(value)) {
+                return value;
+            }
+        }
+        throw new IllegalArgumentException("All values are NaN");
+    }
+
+    @VisibleForTesting
+    JoinComplementStats calculateJoinComplementStats(
+            Optional<Expression> filter,
+            List<JoinNode.EquiJoinClause> criteria,
+            PlanNodeStatsEstimate leftStats,
+            PlanNodeStatsEstimate rightStats)
+    {
+        // TODO: add support for non-equality conditions (e.g: <=, !=, >)
+        if (filter.isPresent()) {
+            // non-equi filters are not supported
+            return new JoinComplementStats(UNKNOWN_STATS, Optional.empty());
+        }
+
+        if (criteria.isEmpty()) {
+            if (rightStats.getOutputRowCount() > 0) {
+                // all left side rows are matched
+                return new JoinComplementStats(leftStats.mapOutputRowCount(rowCount -> 0.0), Optional.empty());
+            }
+            if (rightStats.getOutputRowCount() == 0) {
+                // none left side rows are matched
+                return new JoinComplementStats(leftStats, Optional.empty());
+            }
+            // right stats row count is NaN
+            return new JoinComplementStats(UNKNOWN_STATS, Optional.empty());
+        }
+
+        // heuristics: select the most selective criteria for join complement clause
+        return IntStream.range(0, criteria.size())
+                .mapToObj(drivingClauseId -> {
+                    EquiJoinClause drivingClause = criteria.get(drivingClauseId);
+                    List<EquiJoinClause> remainingClauses = copyWithout(criteria, drivingClauseId);
+                    return calculateJoinComplementStats(leftStats, rightStats, drivingClause, remainingClauses);
+                })
+                .max(comparingDouble(joinComplementStats -> joinComplementStats.getStats().getOutputRowCount()))
+                .get();
+    }
+
+    private JoinComplementStats calculateJoinComplementStats(
+            PlanNodeStatsEstimate leftStats,
+            PlanNodeStatsEstimate rightStats,
+            EquiJoinClause drivingClause,
+            List<EquiJoinClause> remainingClauses)
+    {
+        PlanNodeStatsEstimate result = leftStats;
+
+        SymbolStatsEstimate leftColumnStats = leftStats.getSymbolStatistics(drivingClause.getLeft());
+        SymbolStatsEstimate rightColumnStats = rightStats.getSymbolStatistics(drivingClause.getRight());
+
+        // TODO: use range methods when they have defined (and consistent) semantics
+        double leftNDV = leftColumnStats.getDistinctValuesCount();
+        double matchingRightNDV = rightColumnStats.getDistinctValuesCount() * unmatchedJoinComplementNdvsCoefficient;
+
+        if (leftNDV > matchingRightNDV) {
+            double drivingClauseRangeFraction = leftColumnStats.getValuesFraction() * (leftNDV - matchingRightNDV) / leftNDV;
```

Comment:
I don't know what rangeFraction means.  Maybe call this `nonMatchingLeftValuesFraction`?  it's sort of clunky but I think easier to understand

Reply:
@sopel39 ?